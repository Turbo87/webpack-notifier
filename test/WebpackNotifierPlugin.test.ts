import {join} from 'path';
import {
  contentImageSerializer,
  reduceArraySerializer,
  changedOptionsSerializer,
  testChangesFlow,
  TestArguments,
  skipNotifierCallback
} from './helpers/utils';
import CustomWarningPlugin from './helpers/CustomWarningPlugin';
import ChildCompilationPlugin from './helpers/ChildCompilationPlugin';

expect.addSnapshotSerializer(reduceArraySerializer);
expect.addSnapshotSerializer(contentImageSerializer);
expect.addSnapshotSerializer(skipNotifierCallback);
expect.addSnapshotSerializer(changedOptionsSerializer);

describe('WebpackNotifierPlugin', () => {
  describe('one compilation', () => {
    test.each([
      [['successful'], undefined],
      [['error'], undefined],
      [['warning'], undefined],
    ])('%j %j', testChangesFlow);
  });
  describe('title', () => {
    test.each([
      [['successful'], {title: 'Webpack'}],
      [['successful'], {title}],
      [['error'], {title}],
      [['warning'], {title}],
    ])('%j %j', testChangesFlow);
    function title({msg}) {
      if (msg.startsWith('Error')) return 'build error ❌';
      if (msg.startsWith('Warning')) return 'build warning ⚠️';
      return 'build complete ✅';
    }
    describe('new title function API', () => {
      test.each([
        [['successful'], {title: 'Webpack'}],
        [['successful'], {title}],
        [['error'], {title}],
        [['warning'], {title}],
      ])('%j %j', testChangesFlow);
      function title(params) {
        return `Build status is ${params.status} with message ${params.message}`;
      }
    });
  });
  describe('emoji message', () => {
    test.each<TestArguments>([
      [['successful'], {emoji: true}],
      [['error'], {emoji: true}],
      [['warning'], {emoji: true}],
      [['successful'], {emoji: true}, {plugins: [new CustomWarningPlugin()]}],
    ])('%j %j %j', testChangesFlow);
  });
  describe('contentImage', () => {
    const contentImage = {
      success: join(__dirname, '../successImage.png'),
      warning: join(__dirname, '../warningsImage.png'),
      error: join(__dirname, '../errorsImage.png')
    };
    test.each([
      [['successful'], {
        contentImage: join(__dirname, '../another-logo.png')
      }],
    ])('%j {contentImage: "../another-logo.png"}', testChangesFlow);
    test.each([
      [['successful'], {contentImage}],
      [['error'], {contentImage}],
      [['warning'], {contentImage}],
    ])('%j {contentImage: {success: "../successImage.png"}, error: "../errorImage.png"}, warning: "../warningImage.png"}}', testChangesFlow);
  });
  describe('verbosity level configuration', () => {
    describe('Default', () => {
      test.each([
        [['successful', 'successful', 'successful'], undefined],
        [['error', 'error', 'successful'], undefined],
      ])('%j %j', testChangesFlow);
    });
    describe('Exclude Warnings', () => {
      test.each([
        [['warning'], {excludeWarnings: true}],
      ])('%j %j', testChangesFlow);
    });
    describe('Always Notify', () => {
      test.each([
        [['successful', 'successful'], {alwaysNotify: true}],
      ])('%j %j', testChangesFlow);
    });
    describe('Notify on error', () => {
      test.each([
        [['successful', 'warning', 'error'], {onlyOnError: true}],
      ])('%j %j', testChangesFlow);
    });
    describe('Skip Notification on the First Build', () => {
      test.each([
        [['successful', 'successful'], {skipFirstNotification: true}],
      ])('%j %j', testChangesFlow);
    });
  });

  describe('custom warning', () => {// TODO maybe deprecated
    test.each([
      [['successful'], undefined, {plugins: [new CustomWarningPlugin()]}],
    ])('%j %j', testChangesFlow);
  });
  describe('child compilation errors', () => {
    test.each([
      [['successful'], undefined, {plugins: [new ChildCompilationPlugin('Warning')]}],
      [['successful'], undefined, {plugins: [new ChildCompilationPlugin('Error')]}],
      [['successful'], undefined, {plugins: [new ChildCompilationPlugin(), new ChildCompilationPlugin('Warning')]}],
    ])('%j %j %j', testChangesFlow, 10e3);
  });
});
