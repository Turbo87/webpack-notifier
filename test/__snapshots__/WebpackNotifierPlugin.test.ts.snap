// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebpackNotifierPlugin child compilation errors ["successful"] undefined {"plugins":[{"level":"Error","levelCollectionKey":"errors"}]}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": undefined,
    "line": undefined,
  },
  "message": "Warning: Child Compilation Error",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin child compilation errors ["successful"] undefined {"plugins":[{"level":"Warning","levelCollectionKey":"warnings"}]}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": undefined,
    "line": undefined,
  },
  "message": "Warning: Child Compilation Warning",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin child compilation errors ["successful"] undefined {"plugins":[{"level":false},{"level":"Warning","levelCollectionKey":"warnings"}]}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": undefined,
    "line": undefined,
  },
  "message": "Warning: Child Compilation Warning",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin contentImage ["error"] {contentImage: {success: "../successImage.png"}, error: "../errorImage.png"}, warning: "../warningImage.png"}}: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/errorsImage.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin contentImage ["successful"] {contentImage: "../another-logo.png"}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/another-logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin contentImage ["successful"] {contentImage: {success: "../successImage.png"}, error: "../errorImage.png"}, warning: "../warningImage.png"}}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/successImage.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin contentImage ["warning"] {contentImage: {success: "../successImage.png"}, error: "../errorImage.png"}, warning: "../warningImage.png"}}: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/warningsImage.png",
  "location": Object {
    "column": undefined,
    "file": "/entry.js",
    "line": undefined,
  },
  "message": "Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin custom warning ["successful"] undefined: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": undefined,
    "line": undefined,
  },
  "message": "Warning: Custom Warning message",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin emoji message ["error"] {"emoji":true} %j: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "emoji": true,
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "❌ Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin emoji message ["successful"] {"emoji":true} %j: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "emoji": true,
  "location": undefined,
  "message": "✅ Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin emoji message ["successful"] {"emoji":true} {"plugins":[{}]}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "emoji": true,
  "location": Object {
    "column": undefined,
    "file": undefined,
    "line": undefined,
  },
  "message": "⚠️ Warning: Custom Warning message",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin emoji message ["warning"] {"emoji":true} %j: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "emoji": true,
  "location": Object {
    "column": undefined,
    "file": "/entry.js",
    "line": undefined,
  },
  "message": "⚠️ Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin one compilation ["error"] undefined: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin one compilation ["successful"] undefined: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin one compilation ["warning"] undefined: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": "/entry.js",
    "line": undefined,
  },
  "message": "Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title ["error"] {}: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "build error ❌",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title ["successful"] {"title":"Webpack"}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title ["successful"] {}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "build complete ✅",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title ["warning"] {}: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": "/entry.js",
    "line": undefined,
  },
  "message": "Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "title": "build warning ⚠️",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title new title function API ["error"] {}: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Build status is error with message Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title new title function API ["successful"] {"title":"Webpack"}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title new title function API ["successful"] {}: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Build status is success with message Build successful",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin title new title function API ["warning"] {}: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": undefined,
    "file": "/entry.js",
    "line": undefined,
  },
  "message": "Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "title": "Build status is warning with message Warning: /entry.js
require.extensions is not supported by webpack. Use a loader instead.",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Always Notify ["successful","successful"] {"alwaysNotify":true}: after "successful" build 1`] = `
Object {
  "alwaysNotify": true,
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Always Notify ["successful","successful"] {"alwaysNotify":true}: after "successful" build 2`] = `
Object {
  "alwaysNotify": true,
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["error","error","successful"] undefined: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["error","error","successful"] undefined: after "error" build 2`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["error","error","successful"] undefined: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["successful","successful","successful"] undefined: after "successful" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["successful","successful","successful"] undefined: after "successful" build 2`] = `Array []`;

exports[`WebpackNotifierPlugin verbosity level configuration Default ["successful","successful","successful"] undefined: after "successful" build 3`] = `Array []`;

exports[`WebpackNotifierPlugin verbosity level configuration Exclude Warnings ["warning"] {"excludeWarnings":true}: after "warning" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "excludeWarnings": true,
  "location": undefined,
  "message": "Build successful",
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Notify on error ["successful","warning","error"] {"onlyOnError":true}: after "error" build 1`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": Object {
    "column": 8,
    "file": "/entry.js",
    "line": 1,
  },
  "message": "Error: /entry.js
SyntaxError: Unexpected token (1:8)",
  "onlyOnError": true,
  "title": "Webpack",
  "wait": false,
}
`;

exports[`WebpackNotifierPlugin verbosity level configuration Notify on error ["successful","warning","error"] {"onlyOnError":true}: after "successful" build 1`] = `Array []`;

exports[`WebpackNotifierPlugin verbosity level configuration Notify on error ["successful","warning","error"] {"onlyOnError":true}: after "warning" build 1`] = `Array []`;

exports[`WebpackNotifierPlugin verbosity level configuration Skip Notification on the First Build ["successful","successful"] {"skipFirstNotification":true}: after "successful" build 1`] = `Array []`;

exports[`WebpackNotifierPlugin verbosity level configuration Skip Notification on the First Build ["successful","successful"] {"skipFirstNotification":true}: after "successful" build 2`] = `
Object {
  "contentImage": "__dirname/logo.png",
  "location": undefined,
  "message": "Build successful",
  "skipFirstNotification": true,
  "title": "Webpack",
  "wait": false,
}
`;
